File: cli/js/colors.ts
// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.

interface Code {
  open: string;
  close: string;
  regexp: RegExp;
}

function code(open: number, close: number): Code {
  return {
    open: `\x1b[${open}m`,
    close: `\x1b[${close}m`,
    regexp: new RegExp(`\\x1b\\[${close}m`, "g"),
  };
}

function run(str: string, code: Code): string {
  return !globalThis || !globalThis.Deno || globalThis.Deno.noColor
    ? str
    : `${code.open}${str.replace(code.regexp, code.open)}${code.close}`;
}

export function bold(str: string): string {
  return run(str, code(1, 22));
}

export function italic(str: string): string {
  return run(str, code(3, 23));
}

export function yellow(str: string): string {
  return run(str, code(33, 39));
}

export function cyan(str: string): string {
  return run(str, code(36, 39));
}

export function red(str: string): string {
  return run(str, code(31, 39));
}

export function green(str: string): string {
  return run(str, code(32, 39));
}

export function bgRed(str: string): string {
  return run(str, code(41, 49));
}

export function white(str: string): string {
  return run(str, code(37, 39));
}

export function gray(str: string): string {
  return run(str, code(90, 39));
}

export function magenta(str: string): string {
  return run(str, code(35, 39));
}

export function dim(str: string): string {
  return run(str, code(2, 22));
}

// https://github.com/chalk/ansi-regex/blob/2b56fb0c7a07108e5b54241e8faec160d393aedb/index.js
const ANSI_PATTERN = new RegExp(
  [
    "[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)",
    "(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-ntqry=><~]))",
  ].join("|"),
  "g"
);

export function stripColor(string: string): string {
  return string.replace(ANSI_PATTERN, "");
}

File: cli/colors.rs
// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
use regex::Regex;
use std::env;
use std::fmt;
use std::io::Write;
use termcolor::Color::{Ansi256, Black, Magenta, Red, White};
use termcolor::{Ansi, ColorSpec, WriteColor};

#[cfg(windows)]
use termcolor::{BufferWriter, ColorChoice};

lazy_static! {
        // STRIP_ANSI_RE and strip_ansi_codes are lifted from the "console" crate.
        // Copyright 2017 Armin Ronacher <armin.ronacher@active-4.com>. MIT License.
        static ref STRIP_ANSI_RE: Regex = Regex::new(
                r"[\x1b\x9b][\[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-PRZcf-nqry=><]"
        ).unwrap();
        static ref NO_COLOR: bool = {
                env::var_os("NO_COLOR").is_some()
        };
}

/// Helper function to strip ansi codes.
pub fn strip_ansi_codes(s: &str) -> std::borrow::Cow<str> {
  STRIP_ANSI_RE.replace_all(s, "")
}

pub fn use_color() -> bool {
  !(*NO_COLOR)
}

#[cfg(windows)]
pub fn enable_ansi() {
  BufferWriter::stdout(ColorChoice::AlwaysAnsi);
}

fn style(s: &str, colorspec: ColorSpec) -> impl fmt::Display {
  if !use_color() {
    return String::from(s);
  }
  let mut v = Vec::new();
  let mut ansi_writer = Ansi::new(&mut v);
  ansi_writer.set_color(&colorspec).unwrap();
  ansi_writer.write_all(s.as_bytes()).unwrap();
  ansi_writer.reset().unwrap();
  String::from_utf8_lossy(&v).into_owned()
}

pub fn red_bold(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Red)).set_bold(true);
  style(&s, style_spec)
}

pub fn green_bold(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Ansi256(10))).set_bold(true);
  style(&s, style_spec)
}

pub fn italic_bold(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_bold(true).set_italic(true);
  style(&s, style_spec)
}

pub fn black_on_white(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_bg(Some(White)).set_fg(Some(Black));
  style(&s, style_spec)
}

pub fn white_on_red(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_bg(Some(Red)).set_fg(Some(White));
  style(&s, style_spec)
}

pub fn white_on_green(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_bg(Some(Ansi256(10))).set_fg(Some(White));
  style(&s, style_spec)
}

pub fn yellow(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Ansi256(11)));
  style(&s, style_spec)
}

pub fn cyan(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Ansi256(14)));
  style(&s, style_spec)
}

pub fn red(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Red));
  style(&s, style_spec)
}

pub fn green(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Ansi256(10)));
  style(&s, style_spec)
}

pub fn magenta(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Magenta));
  style(&s, style_spec)
}

pub fn bold(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_bold(true);
  style(&s, style_spec)
}

pub fn gray(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec.set_fg(Some(Ansi256(8)));
  style(&s, style_spec)
}

pub fn italic_bold_gray(s: String) -> impl fmt::Display {
  let mut style_spec = ColorSpec::new();
  style_spec
    .set_fg(Some(Ansi256(8)))
    .set_bold(true)
    .set_italic(true);
  style(&s, style_spec)
}
